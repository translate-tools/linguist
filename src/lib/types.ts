import { isRight } from 'fp-ts/lib/Either';
import * as t from 'io-ts';

export const type = t;

// TODO: remove not declared properties
/**
 * Try to decode object and return object with data or with errors
 */
export function decodeStruct<T>(
	type: t.Type<T>,
	data: any,
):
	| { data: T; errors: null }
	| {
			data: null;
			errors: {
				key: string;
				value: unknown;
				type: t.Decoder<any, any>;
				message?: string;
			}[];
	  } {
	const decodeResult = type.decode(data);

	// Return decoded value
	if (isRight(decodeResult)) {
		return {
			data: decodeResult.right,
			errors: null,
		};
	}

	// Build errors object
	return {
		data: null,
		errors: decodeResult.left.map((error) => {
			// Remove root object from context tree
			const context = error.context;
			const targetPropertyContext = context[context.length - 1];

			return {
				key: context
					.slice(1)
					.map(({ key }) => key)
					.join('.'),
				value: error.value,
				type: targetPropertyContext.type,
				message: error.message,
			};
		}),
	};
}

// TODO: decode primitive values and generate reports like in `decodeStruct`
/**
 * Helper for decode data by type
 */
export function tryDecode<T>(type: t.Type<T>, data: any): T;
export function tryDecode<T>(type: t.Type<T>, data: any, defaultData: T): T;
export function tryDecode<T>(type: t.Type<T>, data: any, defaultData?: T) {
	const decodedData = type.decode(data);
	if (isRight(decodedData)) {
		return decodedData.right;
	}

	// TODO: remove this
	if (arguments.length >= 3) {
		return defaultData as T;
	}

	console.error('Data for the error below', data);
	throw new TypeError('Invalid type');
}

/**
 * Same as `tryDecode` but only for objects and more verbose
 *
 * @deprecated: use `decodeStruct` instead and generate exception
 */
export const tryDecodeObject = <T extends t.Props>(
	type: t.TypeC<T> | t.PartialC<T>,
	data: any,
) => {
	if (!(data instanceof Object)) {
		throw new TypeError('Data is not object');
	}

	const typeProps = type.props;

	const typeKeys = Object.keys(typeProps);
	const dataKeys = Object.keys(data);
	if (typeKeys.length !== dataKeys.length) {
		throw new RangeError('Number of elements in types and data is not equal');
	}

	for (const key in typeProps) {
		if (!(key in data)) {
			throw new RangeError(`Key "${key}" is not found`);
		}

		const decodedData = typeProps[key].decode(data[key]);
		if (!isRight(decodedData)) {
			throw new TypeError(`Invalid type of key "${key}"`);
		}
	}

	return data as t.TypeOfProps<T>;
};

/**
 * Validate type of value from TypeC object by path
 */
export const checkTypeByPath = <T extends t.Props>(
	typeObject: t.TypeC<T>,
	path: string[],
	value: any,
): boolean => {
	let type: t.TypeC<any> | t.Type<any> = typeObject;

	for (let i = 0; i < path.length; i++) {
		const segment = path[i];
		if ('props' in type) {
			type = type.props[segment];
		} else {
			return false;
		}
	}

	return isRight(type.decode(value));
};

// Type constructors

export const StringLiteralType = <T extends string>(stringOfType: T) =>
	new type.Type(
		`String["${stringOfType}"]`,
		(input: unknown): input is T =>
			typeof input === 'string' && input === stringOfType,
		(input, context) =>
			typeof input === 'string' && input === stringOfType
				? type.success(input as T)
				: type.failure(input, context),
		type.identity,
	);

export const StringPatternType = <T extends string = string>(pattern: RegExp) =>
	new type.Type(
		`StringPatternType["${pattern.source}"]`,
		(input: unknown): input is T => typeof input === 'string' && pattern.test(input),
		(input, context) =>
			typeof input === 'string' && pattern.test(input)
				? type.success(input as T)
				: type.failure(input, context),
		type.identity,
	);
