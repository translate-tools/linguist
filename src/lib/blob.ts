/**
 * Source: https://stackoverflow.com/a/18650249
 */
export const blobToBase64 = (blob: Blob) => {
	return new Promise<string>((resolve) => {
		const reader = new FileReader();
		reader.onloadend = () => {
			if (typeof reader.result === 'string') {
				resolve(reader.result);
			} else {
				throw new TypeError('Unexpected type of result');
			}
		};
		reader.readAsDataURL(blob);
	});
};

/**
 * Source: https://stackoverflow.com/a/16245768
 */
export const base64ToBlob = (b64Data: string, contentType = '', sliceSize = 512) => {
	const byteCharacters = atob(b64Data);
	const byteArrays = [];

	for (let offset = 0; offset < byteCharacters.length; offset += sliceSize) {
		const slice = byteCharacters.slice(offset, offset + sliceSize);

		const byteNumbers = new Array(slice.length);
		for (let i = 0; i < slice.length; i++) {
			byteNumbers[i] = slice.charCodeAt(i);
		}

		const byteArray = new Uint8Array(byteNumbers);
		byteArrays.push(byteArray);
	}

	const blob = new Blob(byteArrays, { type: contentType });
	return blob;
};
