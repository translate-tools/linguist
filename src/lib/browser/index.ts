import browser from 'webextension-polyfill';

import { detectLanguage, isValidLanguage } from '../language';

/**
 * Helper to detect platform
 */
export const isMobileBrowser = () => {
	const UA = navigator.userAgent;
	const isMobileUserAgent =
		/\b(BlackBerry|webOS|iPhone|IEMobile)\b/i.test(UA) ||
		/\b(Android|Windows Phone|iPad|iPod)\b/i.test(UA);

	return isMobileUserAgent;
};

export const injectStyles = (paths: string[], parent?: Node) => {
	paths.forEach((path) => {
		const link = document.createElement('link');
		link.href = browser.runtime.getURL(path);
		link.rel = 'stylesheet';
		(parent !== undefined ? parent : document.head).appendChild(link);
	});
};

export function getPageLanguageFromMeta() {
	const html = document.documentElement;

	const langAttributes = ['lang', 'xml:lang'];
	for (const name of langAttributes) {
		const pageLangRaw = html.getAttribute(name);
		if (pageLangRaw !== null) {
			const match = /^([a-z]+)(-[a-zA-Z]+)?$/.exec(pageLangRaw);
			if (match !== null) {
				const language = match[1];
				return isValidLanguage(language) ? language : null;
			}
		}
	}

	return null;
}

export const isFirefox = () => /firefox/i.test(navigator.userAgent);
export const isChromium = () => /chrome/i.test(navigator.userAgent);
export const isBackgroundContext = () =>
	['background-script.js', '_generated_background_page.html'].some(
		(path) => location.href === browser.runtime.getURL(path),
	);

const extensionHostname = new URL(browser.runtime.getURL('')).host;
export const isExtensionContext = location.host === extensionHostname;

/**
 * By default detect lang by meta, but while `detectByContent` is `true` its try detect lang by content
 */
export const getPageLanguage = async (detectByContent = false, reliableOnly = false) => {
	const langFromMeta = getPageLanguageFromMeta();

	// Try detect language by content
	if (langFromMeta === null || detectByContent) {
		const contentLang = await detectLanguage(document.body.innerText, reliableOnly);
		if (contentLang !== null) {
			return contentLang;
		}
	}

	return langFromMeta;
};
