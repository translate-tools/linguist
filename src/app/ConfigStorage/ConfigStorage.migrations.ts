import browser from 'webextension-polyfill';

import { DEFAULT_TRANSLATOR, DEFAULT_TTS, defaultConfig } from '../../config';
import { createMigrationTask, Migration } from '../../lib/migrations/createMigrationTask';
import { decodeStruct } from '../../lib/types';
import { AppConfig } from '../../types/runtime';

const migrations: Migration[] = [
	{
		version: 1,
		async migrate() {
			const storageKey = 'config.Main';
			const storageDataRaw = localStorage.getItem(storageKey);

			// Skip
			if (storageDataRaw === null) return;

			const storageNameV2 = 'appConfig';

			// Import valid data
			const storageData = JSON.parse(storageDataRaw);
			if (typeof storageData === 'object') {
				// Merge actual data with legacy
				let { [storageNameV2]: actualData } = await browser.storage.local.get(
					storageNameV2,
				);
				if (typeof actualData !== 'object') {
					actualData = {};
				}

				const mergedData = { ...actualData, ...storageData };

				// Write data
				await browser.storage.local.set({
					[storageNameV2]: mergedData,
				});
			}

			// Delete old data
			localStorage.removeItem(storageKey);
		},
	},
	{
		version: 3,
		async migrate() {
			const storageNameV2 = 'appConfig';

			// Merge actual data with old
			let { [storageNameV2]: actualData } = await browser.storage.local.get(
				storageNameV2,
			);
			if (typeof actualData !== 'object') {
				actualData = {};
			}

			const contentscriptPropData =
				actualData?.contentscript?.selectTranslator || {};
			const quickTranslate = actualData?.selectTranslator?.quickTranslate;

			const newData = actualData;
			delete newData.contentscript;

			if (newData.selectTranslator) {
				delete newData.selectTranslator.quickTranslate;
			}

			// Write data
			await browser.storage.local.set({
				[storageNameV2]: {
					...newData,
					selectTranslator: {
						...newData?.selectTranslator,
						...contentscriptPropData,
						mode: quickTranslate
							? 'quickTranslate'
							: newData?.selectTranslator?.mode,
					},
				},
			});
		},
	},
	{
		version: 5,
		async migrate() {
			const storageName = 'appConfig';

			let { [storageName]: actualData } = await browser.storage.local.get(
				storageName,
			);
			if (typeof actualData !== 'object') {
				actualData = {};
			}

			const updatedConfig = {
				ttsModule: DEFAULT_TTS,
				...actualData,
				pageTranslator: {
					enableContextMenu: false,
					toggleTranslationHotkey: null,
					...actualData?.pageTranslator,
				},
			};

			if (actualData.translatorModule === 'BingTranslatorPublic') {
				updatedConfig.translatorModule = DEFAULT_TRANSLATOR;
			}

			// Write data
			await browser.storage.local.set({ [storageName]: updatedConfig });
		},
	},
	{
		// Add history section
		version: 6,
		async migrate() {
			const storageName = 'appConfig';

			let { [storageName]: actualData } = await browser.storage.local.get(
				storageName,
			);
			if (typeof actualData !== 'object') {
				actualData = {};
			}

			const updatedConfig = {
				...actualData,
				history: {
					enabled: true,
				},
			};

			// Write data
			await browser.storage.local.set({ [storageName]: updatedConfig });
		},
	},
	{
		version: 7,
		async migrate() {
			// Empty migration, to bump migration number and to trigger hook for repair config
		},
	},
	{
		version: 8,
		async migrate() {
			const storageName = 'appConfig';

			let { [storageName]: actualData } = await browser.storage.local.get(
				storageName,
			);
			if (typeof actualData !== 'object') {
				actualData = {};
			}

			const updatedConfig = {
				...actualData,
			};

			delete updatedConfig['appIcon'];

			// Write data
			await browser.storage.local.set({ [storageName]: updatedConfig });
		},
	},
];

export const ConfigStorageMigration = createMigrationTask(migrations, {
	onComplete: async () => {
		// Repair config if necessary
		const storageName = 'appConfig';
		const { [storageName]: config } = await browser.storage.local.get(storageName);

		const { errors } = decodeStruct(AppConfig, config);
		if (errors === null) return;

		console.warn('Config object is invalid, fallback to default config', errors);
		await browser.storage.local.set({ [storageName]: defaultConfig });
	},
});
