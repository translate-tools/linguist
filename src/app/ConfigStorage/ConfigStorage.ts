import { createEvent, createStore, Store } from 'effector';
import browser from 'webextension-polyfill';

import { updateNotEqualProps } from '../../lib/effector/reducers';
import { decodeStruct } from '../../lib/types';
import { AppConfig, AppConfigType } from '../../types/runtime';

export interface AsyncStorage<T> {
	get(): Promise<T>;
	set(data: T): Promise<void>;
}

export class ConfigStorage implements AsyncStorage<AppConfigType> {
	private readonly storageName = 'appConfig';
	private readonly defaultData: AppConfigType;

	constructor(defaultData: AppConfigType) {
		this.defaultData = defaultData;
	}

	public async get() {
		// Get data from storage if possible
		const { [this.storageName]: data } = await browser.storage.local.get(
			this.storageName,
		);

		// Return default data for empty storage
		if (data === undefined) {
			return this.defaultData;
		}

		const configCodec = decodeStruct(AppConfig, data);
		if (configCodec.errors !== null) {
			throw new Error('Invalid config');
		}

		return configCodec.data;
	}

	public async set(data: AppConfigType) {
		await browser.storage.local.set({ [this.storageName]: data });
	}
}

export class ObservableAsyncStorage<T extends Record<any, any>>
	implements AsyncStorage<T>
{
	private readonly config: AsyncStorage<T>;

	constructor(config: AsyncStorage<T>) {
		this.config = config;
	}

	private store: Store<T> | null = null;
	private readonly updateData = createEvent<T>();

	public async getObservableStore() {
		if (this.store === null) {
			const state = await this.config.get();
			this.store = createStore<T>(state);
			this.store.on(this.updateData, updateNotEqualProps);
		}

		return this.store;
	}

	public async get() {
		return this.config.get();
	}

	public async set(data: T) {
		const newObject = { ...data };

		await this.config.set(newObject);
		this.updateData(newObject);
	}
}
