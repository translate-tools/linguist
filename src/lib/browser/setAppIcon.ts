import browser from 'webextension-polyfill';
import { isFirefox } from '@react-aria/utils';

import { AppConfigType } from '../../types/runtime';

export type AppIcon = Exclude<AppConfigType['appIcon'], 'auto'>;

// TODO: remove icon object usage
export const getAppIconPath = (_icon: AppIcon, absolutePath = false) => {
	const relativePath = 'static/logo.png';
	return absolutePath ? browser.runtime.getURL(relativePath) : relativePath;
};

/**
 * Set static icon
 */
export const setAppIcon = (icon: AppIcon) => {
	const iconPath = getAppIconPath(icon);
	(isFirefox() ? browser.browserAction : browser.action).setIcon({
		path: {
			128: iconPath,
		},
	});
};
