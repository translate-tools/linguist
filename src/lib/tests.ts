import { IDBFactory } from 'fake-indexeddb';
import browser from 'webextension-polyfill';

export const wipeIDB = () => {
	// For test purposes
	// eslint-disable-next-line no-global-assign
	indexedDB = new IDBFactory();
};

export const clearAllMocks = async () => {
	wipeIDB();
	localStorage.clear();
	await browser.storage.local.clear();
};
