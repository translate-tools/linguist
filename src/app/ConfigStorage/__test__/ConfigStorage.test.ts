import browser from 'webextension-polyfill';

import { clearAllMocks } from '../../../lib/tests';
import { AppConfigType } from '../../../types/runtime';

import { ConfigStorage, ObservableAsyncStorage } from '../ConfigStorage';
import { ConfigStorageMigration } from '../ConfigStorage.migrations';
import configVersion1 from './config-v1.json';
import configVersion3 from './config-v3.json';
import latestConfigSample from './config-v8.json';

const latestVersion = 8;

describe('config migrations', () => {
	beforeAll(clearAllMocks);

	test('migrate config from v0 to v3', async () => {
		// Load data
		localStorage.setItem('config.Main', JSON.stringify(configVersion1));

		// Migrate data
		await ConfigStorageMigration.migrate(0, 3);

		const { appConfig } = await browser.storage.local.get('appConfig');

		expect(appConfig).toEqual(configVersion3);
		expect(localStorage.getItem('config.Main')).toBeNull();
	});

	test('migrate config v0 to latest version', async () => {
		// Load data
		localStorage.setItem(
			'config.Main',
			JSON.stringify({
				...configVersion1,
				translatorModule: 'BingTranslatorPublic',
			}),
		);

		// Migrate data
		await ConfigStorageMigration.migrate(0, latestVersion);

		const { appConfig } = await browser.storage.local.get('appConfig');
		expect(appConfig).toEqual(latestConfigSample);
	});

	describe(`race condition detection`, () => {
		beforeEach(clearAllMocks);

		for (let attempt = 1; attempt <= 5; attempt++) {
			test(`Detection race conditions. Attempt #${attempt}`, async () => {
				// Load data
				localStorage.setItem(
					'config.Main',
					JSON.stringify({
						...configVersion1,
						translatorModule: 'BingTranslatorPublic',
					}),
				);

				// Migrate part of data
				await ConfigStorageMigration.migrate(0, 5);

				// Migrate another part of data
				await ConfigStorageMigration.migrate(5, latestVersion);

				const { appConfig } = await browser.storage.local.get('appConfig');
				expect(appConfig).toEqual(latestConfigSample);
			});
		}
	});
});

describe('use config', () => {
	beforeAll(clearAllMocks);

	const latestConfigObject = latestConfigSample as AppConfigType;

	test('config storage set/get', async () => {
		const configStorage = new ConfigStorage(latestConfigObject);

		// Get config
		const config1 = await configStorage.get();
		expect(config1).toEqual(latestConfigObject);

		const newData = { ...config1, translatorModule: 'testTranslator' };
		await configStorage.set(newData);

		const config2 = await configStorage.get();
		expect(config2).toEqual(newData);
	});

	test.skip('observable storage', async () => {
		const configStorage = new ConfigStorage(latestConfigObject);
		const observableConfigStorage = new ObservableAsyncStorage(configStorage);

		// Listen config update
		const $config = await observableConfigStorage.getObservableStore();
		const updateConfigPromise = new Promise<AppConfigType>((res) => {
			$config.updates.watch(res);
		});

		const latestConfig = await configStorage.get();
		await observableConfigStorage.set({ ...latestConfig, language: 'ja' });

		const updatedConfig = await updateConfigPromise;
		expect(updatedConfig.language).toBe('ja');
	});
});
