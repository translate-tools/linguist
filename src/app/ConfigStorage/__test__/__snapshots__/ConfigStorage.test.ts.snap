// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config migrations > migrate config v0 to latest version 1`] = `
{
  "cache": {
    "ignoreCase": true,
  },
  "history": {
    "enabled": true,
  },
  "language": "en",
  "pageTranslator": {
    "detectLanguageByContent": true,
    "enableContextMenu": false,
    "excludeSelectors": [
      "! Basic tags",
      "meta",
      "link",
      "script",
      "noscript",
      "style",
      "code",
      "pre",
      "textarea",
      "time",
      "kbd",
      "svg",
      "g",
      "ruby",
      "rt",
      "rp",
      "math",
      "d-math",
      "samp",
      "",
      "! Elements that request to disable translation",
      ".notranslate",
      "[contenteditable="true"]",
      "[translate=no]",
      "",
      "! Social sharing buttons",
      ".social-share",
      ".share-nav",
      "[data-toolbar=share]",
      ".o-share",
      "",
      "! Code elements",
      ".prism-code",
      ".enlighter-code",
      ".rc-CodeBlock",
      "[role=code]",
      "table.highlight",
      "",
      "! example: https://web.hypothes.is/blog/do-it-yourself-anchoring-and-the-evolution-of-the-hypothesis-toolkit/",
      "hypothesis-highlight",
      ".hypothesis-highlight",
      "",
      "! Icons",
      ".material-icons",
      "material-icon",
      "span[class^=material-symbols-]",
      ".google-symbols",
      "i.fa",
      "i[class^=fa-]",
      "",
      "! Other",
      "visuallyhidden",
    ],
    "lazyTranslate": true,
    "originalTextPopup": false,
    "toggleTranslationHotkey": null,
    "translatableAttributes": [
      "title",
      "alt",
      "placeholder",
      "label",
      "aria-label",
      "value",
    ],
  },
  "popup": {
    "rememberLastTab": true,
  },
  "popupTab": {
    "pageTranslator": {
      "showCounters": true,
    },
  },
  "scheduler": {
    "chunkSizeForInstantTranslate": null,
    "directTranslateLength": null,
    "isAllowDirectTranslateBadChunks": true,
    "translatePoolDelay": 300,
    "translateRetryAttemptLimit": 2,
    "useCache": true,
  },
  "selectTranslator": {
    "detectedLangFirst": true,
    "disableWhileTranslatePage": false,
    "enabled": true,
    "focusOnTranslateButton": false,
    "isUseAutoForDetectLang": true,
    "mode": "quickTranslate",
    "modifiers": [],
    "rememberDirection": false,
    "showOnceForSelection": true,
    "showOriginalText": true,
    "strictSelection": false,
    "timeoutForHideButton": 3000,
    "zIndex": 999999,
  },
  "textTranslator": {
    "rememberText": true,
    "spellCheck": true,
    "suggestLanguage": true,
    "suggestLanguageAlways": true,
  },
  "translatorModule": "MicrosoftTranslator",
  "ttsModule": "google",
}
`;

exports[`config migrations > race condition detection > Detection race conditions. Attempt #1 1`] = `
{
  "cache": {
    "ignoreCase": true,
  },
  "history": {
    "enabled": true,
  },
  "language": "en",
  "pageTranslator": {
    "detectLanguageByContent": true,
    "enableContextMenu": false,
    "excludeSelectors": [
      "! Basic tags",
      "meta",
      "link",
      "script",
      "noscript",
      "style",
      "code",
      "pre",
      "textarea",
      "time",
      "kbd",
      "svg",
      "g",
      "ruby",
      "rt",
      "rp",
      "math",
      "d-math",
      "samp",
      "",
      "! Elements that request to disable translation",
      ".notranslate",
      "[contenteditable="true"]",
      "[translate=no]",
      "",
      "! Social sharing buttons",
      ".social-share",
      ".share-nav",
      "[data-toolbar=share]",
      ".o-share",
      "",
      "! Code elements",
      ".prism-code",
      ".enlighter-code",
      ".rc-CodeBlock",
      "[role=code]",
      "table.highlight",
      "",
      "! example: https://web.hypothes.is/blog/do-it-yourself-anchoring-and-the-evolution-of-the-hypothesis-toolkit/",
      "hypothesis-highlight",
      ".hypothesis-highlight",
      "",
      "! Icons",
      ".material-icons",
      "material-icon",
      "span[class^=material-symbols-]",
      ".google-symbols",
      "i.fa",
      "i[class^=fa-]",
      "",
      "! Other",
      "visuallyhidden",
    ],
    "lazyTranslate": true,
    "originalTextPopup": false,
    "toggleTranslationHotkey": null,
    "translatableAttributes": [
      "title",
      "alt",
      "placeholder",
      "label",
      "aria-label",
      "value",
    ],
  },
  "popup": {
    "rememberLastTab": true,
  },
  "popupTab": {
    "pageTranslator": {
      "showCounters": true,
    },
  },
  "scheduler": {
    "chunkSizeForInstantTranslate": null,
    "directTranslateLength": null,
    "isAllowDirectTranslateBadChunks": true,
    "translatePoolDelay": 300,
    "translateRetryAttemptLimit": 2,
    "useCache": true,
  },
  "selectTranslator": {
    "detectedLangFirst": true,
    "disableWhileTranslatePage": false,
    "enabled": true,
    "focusOnTranslateButton": false,
    "isUseAutoForDetectLang": true,
    "mode": "quickTranslate",
    "modifiers": [],
    "rememberDirection": false,
    "showOnceForSelection": true,
    "showOriginalText": true,
    "strictSelection": false,
    "timeoutForHideButton": 3000,
    "zIndex": 999999,
  },
  "textTranslator": {
    "rememberText": true,
    "spellCheck": true,
    "suggestLanguage": true,
    "suggestLanguageAlways": true,
  },
  "translatorModule": "MicrosoftTranslator",
  "ttsModule": "google",
}
`;

exports[`config migrations > race condition detection > Detection race conditions. Attempt #2 1`] = `
{
  "cache": {
    "ignoreCase": true,
  },
  "history": {
    "enabled": true,
  },
  "language": "en",
  "pageTranslator": {
    "detectLanguageByContent": true,
    "enableContextMenu": false,
    "excludeSelectors": [
      "! Basic tags",
      "meta",
      "link",
      "script",
      "noscript",
      "style",
      "code",
      "pre",
      "textarea",
      "time",
      "kbd",
      "svg",
      "g",
      "ruby",
      "rt",
      "rp",
      "math",
      "d-math",
      "samp",
      "",
      "! Elements that request to disable translation",
      ".notranslate",
      "[contenteditable="true"]",
      "[translate=no]",
      "",
      "! Social sharing buttons",
      ".social-share",
      ".share-nav",
      "[data-toolbar=share]",
      ".o-share",
      "",
      "! Code elements",
      ".prism-code",
      ".enlighter-code",
      ".rc-CodeBlock",
      "[role=code]",
      "table.highlight",
      "",
      "! example: https://web.hypothes.is/blog/do-it-yourself-anchoring-and-the-evolution-of-the-hypothesis-toolkit/",
      "hypothesis-highlight",
      ".hypothesis-highlight",
      "",
      "! Icons",
      ".material-icons",
      "material-icon",
      "span[class^=material-symbols-]",
      ".google-symbols",
      "i.fa",
      "i[class^=fa-]",
      "",
      "! Other",
      "visuallyhidden",
    ],
    "lazyTranslate": true,
    "originalTextPopup": false,
    "toggleTranslationHotkey": null,
    "translatableAttributes": [
      "title",
      "alt",
      "placeholder",
      "label",
      "aria-label",
      "value",
    ],
  },
  "popup": {
    "rememberLastTab": true,
  },
  "popupTab": {
    "pageTranslator": {
      "showCounters": true,
    },
  },
  "scheduler": {
    "chunkSizeForInstantTranslate": null,
    "directTranslateLength": null,
    "isAllowDirectTranslateBadChunks": true,
    "translatePoolDelay": 300,
    "translateRetryAttemptLimit": 2,
    "useCache": true,
  },
  "selectTranslator": {
    "detectedLangFirst": true,
    "disableWhileTranslatePage": false,
    "enabled": true,
    "focusOnTranslateButton": false,
    "isUseAutoForDetectLang": true,
    "mode": "quickTranslate",
    "modifiers": [],
    "rememberDirection": false,
    "showOnceForSelection": true,
    "showOriginalText": true,
    "strictSelection": false,
    "timeoutForHideButton": 3000,
    "zIndex": 999999,
  },
  "textTranslator": {
    "rememberText": true,
    "spellCheck": true,
    "suggestLanguage": true,
    "suggestLanguageAlways": true,
  },
  "translatorModule": "MicrosoftTranslator",
  "ttsModule": "google",
}
`;

exports[`config migrations > race condition detection > Detection race conditions. Attempt #3 1`] = `
{
  "cache": {
    "ignoreCase": true,
  },
  "history": {
    "enabled": true,
  },
  "language": "en",
  "pageTranslator": {
    "detectLanguageByContent": true,
    "enableContextMenu": false,
    "excludeSelectors": [
      "! Basic tags",
      "meta",
      "link",
      "script",
      "noscript",
      "style",
      "code",
      "pre",
      "textarea",
      "time",
      "kbd",
      "svg",
      "g",
      "ruby",
      "rt",
      "rp",
      "math",
      "d-math",
      "samp",
      "",
      "! Elements that request to disable translation",
      ".notranslate",
      "[contenteditable="true"]",
      "[translate=no]",
      "",
      "! Social sharing buttons",
      ".social-share",
      ".share-nav",
      "[data-toolbar=share]",
      ".o-share",
      "",
      "! Code elements",
      ".prism-code",
      ".enlighter-code",
      ".rc-CodeBlock",
      "[role=code]",
      "table.highlight",
      "",
      "! example: https://web.hypothes.is/blog/do-it-yourself-anchoring-and-the-evolution-of-the-hypothesis-toolkit/",
      "hypothesis-highlight",
      ".hypothesis-highlight",
      "",
      "! Icons",
      ".material-icons",
      "material-icon",
      "span[class^=material-symbols-]",
      ".google-symbols",
      "i.fa",
      "i[class^=fa-]",
      "",
      "! Other",
      "visuallyhidden",
    ],
    "lazyTranslate": true,
    "originalTextPopup": false,
    "toggleTranslationHotkey": null,
    "translatableAttributes": [
      "title",
      "alt",
      "placeholder",
      "label",
      "aria-label",
      "value",
    ],
  },
  "popup": {
    "rememberLastTab": true,
  },
  "popupTab": {
    "pageTranslator": {
      "showCounters": true,
    },
  },
  "scheduler": {
    "chunkSizeForInstantTranslate": null,
    "directTranslateLength": null,
    "isAllowDirectTranslateBadChunks": true,
    "translatePoolDelay": 300,
    "translateRetryAttemptLimit": 2,
    "useCache": true,
  },
  "selectTranslator": {
    "detectedLangFirst": true,
    "disableWhileTranslatePage": false,
    "enabled": true,
    "focusOnTranslateButton": false,
    "isUseAutoForDetectLang": true,
    "mode": "quickTranslate",
    "modifiers": [],
    "rememberDirection": false,
    "showOnceForSelection": true,
    "showOriginalText": true,
    "strictSelection": false,
    "timeoutForHideButton": 3000,
    "zIndex": 999999,
  },
  "textTranslator": {
    "rememberText": true,
    "spellCheck": true,
    "suggestLanguage": true,
    "suggestLanguageAlways": true,
  },
  "translatorModule": "MicrosoftTranslator",
  "ttsModule": "google",
}
`;

exports[`config migrations > race condition detection > Detection race conditions. Attempt #4 1`] = `
{
  "cache": {
    "ignoreCase": true,
  },
  "history": {
    "enabled": true,
  },
  "language": "en",
  "pageTranslator": {
    "detectLanguageByContent": true,
    "enableContextMenu": false,
    "excludeSelectors": [
      "! Basic tags",
      "meta",
      "link",
      "script",
      "noscript",
      "style",
      "code",
      "pre",
      "textarea",
      "time",
      "kbd",
      "svg",
      "g",
      "ruby",
      "rt",
      "rp",
      "math",
      "d-math",
      "samp",
      "",
      "! Elements that request to disable translation",
      ".notranslate",
      "[contenteditable="true"]",
      "[translate=no]",
      "",
      "! Social sharing buttons",
      ".social-share",
      ".share-nav",
      "[data-toolbar=share]",
      ".o-share",
      "",
      "! Code elements",
      ".prism-code",
      ".enlighter-code",
      ".rc-CodeBlock",
      "[role=code]",
      "table.highlight",
      "",
      "! example: https://web.hypothes.is/blog/do-it-yourself-anchoring-and-the-evolution-of-the-hypothesis-toolkit/",
      "hypothesis-highlight",
      ".hypothesis-highlight",
      "",
      "! Icons",
      ".material-icons",
      "material-icon",
      "span[class^=material-symbols-]",
      ".google-symbols",
      "i.fa",
      "i[class^=fa-]",
      "",
      "! Other",
      "visuallyhidden",
    ],
    "lazyTranslate": true,
    "originalTextPopup": false,
    "toggleTranslationHotkey": null,
    "translatableAttributes": [
      "title",
      "alt",
      "placeholder",
      "label",
      "aria-label",
      "value",
    ],
  },
  "popup": {
    "rememberLastTab": true,
  },
  "popupTab": {
    "pageTranslator": {
      "showCounters": true,
    },
  },
  "scheduler": {
    "chunkSizeForInstantTranslate": null,
    "directTranslateLength": null,
    "isAllowDirectTranslateBadChunks": true,
    "translatePoolDelay": 300,
    "translateRetryAttemptLimit": 2,
    "useCache": true,
  },
  "selectTranslator": {
    "detectedLangFirst": true,
    "disableWhileTranslatePage": false,
    "enabled": true,
    "focusOnTranslateButton": false,
    "isUseAutoForDetectLang": true,
    "mode": "quickTranslate",
    "modifiers": [],
    "rememberDirection": false,
    "showOnceForSelection": true,
    "showOriginalText": true,
    "strictSelection": false,
    "timeoutForHideButton": 3000,
    "zIndex": 999999,
  },
  "textTranslator": {
    "rememberText": true,
    "spellCheck": true,
    "suggestLanguage": true,
    "suggestLanguageAlways": true,
  },
  "translatorModule": "MicrosoftTranslator",
  "ttsModule": "google",
}
`;

exports[`config migrations > race condition detection > Detection race conditions. Attempt #5 1`] = `
{
  "cache": {
    "ignoreCase": true,
  },
  "history": {
    "enabled": true,
  },
  "language": "en",
  "pageTranslator": {
    "detectLanguageByContent": true,
    "enableContextMenu": false,
    "excludeSelectors": [
      "! Basic tags",
      "meta",
      "link",
      "script",
      "noscript",
      "style",
      "code",
      "pre",
      "textarea",
      "time",
      "kbd",
      "svg",
      "g",
      "ruby",
      "rt",
      "rp",
      "math",
      "d-math",
      "samp",
      "",
      "! Elements that request to disable translation",
      ".notranslate",
      "[contenteditable="true"]",
      "[translate=no]",
      "",
      "! Social sharing buttons",
      ".social-share",
      ".share-nav",
      "[data-toolbar=share]",
      ".o-share",
      "",
      "! Code elements",
      ".prism-code",
      ".enlighter-code",
      ".rc-CodeBlock",
      "[role=code]",
      "table.highlight",
      "",
      "! example: https://web.hypothes.is/blog/do-it-yourself-anchoring-and-the-evolution-of-the-hypothesis-toolkit/",
      "hypothesis-highlight",
      ".hypothesis-highlight",
      "",
      "! Icons",
      ".material-icons",
      "material-icon",
      "span[class^=material-symbols-]",
      ".google-symbols",
      "i.fa",
      "i[class^=fa-]",
      "",
      "! Other",
      "visuallyhidden",
    ],
    "lazyTranslate": true,
    "originalTextPopup": false,
    "toggleTranslationHotkey": null,
    "translatableAttributes": [
      "title",
      "alt",
      "placeholder",
      "label",
      "aria-label",
      "value",
    ],
  },
  "popup": {
    "rememberLastTab": true,
  },
  "popupTab": {
    "pageTranslator": {
      "showCounters": true,
    },
  },
  "scheduler": {
    "chunkSizeForInstantTranslate": null,
    "directTranslateLength": null,
    "isAllowDirectTranslateBadChunks": true,
    "translatePoolDelay": 300,
    "translateRetryAttemptLimit": 2,
    "useCache": true,
  },
  "selectTranslator": {
    "detectedLangFirst": true,
    "disableWhileTranslatePage": false,
    "enabled": true,
    "focusOnTranslateButton": false,
    "isUseAutoForDetectLang": true,
    "mode": "quickTranslate",
    "modifiers": [],
    "rememberDirection": false,
    "showOnceForSelection": true,
    "showOriginalText": true,
    "strictSelection": false,
    "timeoutForHideButton": 3000,
    "zIndex": 999999,
  },
  "textTranslator": {
    "rememberText": true,
    "spellCheck": true,
    "suggestLanguage": true,
    "suggestLanguageAlways": true,
  },
  "translatorModule": "MicrosoftTranslator",
  "ttsModule": "google",
}
`;
